### PR TITLE
get filtering box on left hand Library Pane to work!

### PR DESCRIPTION
**Pull request by ahsu1230**
can filter out libraries as you type in letters in the input box! woohoo!
Main & Secret libraries never show up under the search box too

@andrewconner @yipingliao

### CODE REVIEW

File: kifi-backend/modules/shoebox/angular/src/layout/nav/nav.js [hunk 1]
```diff
@@ -34,22 +34,6 @@ angular.module('kifi')
               scope.secretLib = _.find(libraries, function (lib) {
                   return lib.kind === 'system_secret';
               });
-              scope.userLibs = _.filter(libraries, function (lib) {
-                return lib.kind === 'user_created';
-              });
-              scope.invitedLibs = libraryService.invitedSummaries;
-
-              var lines;
```
Comment (ahsu1230) [inline kifi-backend/modules/shoebox/angular/src/layout/nav/nav.js:42]: naming wrapping & stuff moved to libraryService so the firstLine / secondLine fields are added right when libraries are retrieved from servers. No need to do it again anywhere else.

**Comment by andrewconner** [comment]
Merging for demo, will go through more soon.

**Comment by ahsu1230** [comment]
thanks

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 2]
```diff
@@ -192,6 +243,23 @@ angular.module('kifi')
...
+
+      filterLibraries: function(term) {
+        var newMyLibs = librarySummaries;
+        var newMyInvited = invitedSummaries;
+        if (term.length) {
+          newMyLibs = fuseSearch1.search(term);
+          newMyInvited = fuseSearch2.search(term);
+        }
+        newMyLibs = _.filter(newMyLibs, function (lib) {
+          return lib.kind === 'user_created';
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:256]: This filtering to get to the user created library happens in a couple of places. Maybe it'd be convenient to have a userCreatedLibrarySummaries that is created whenever we update librarySummaries, and the rest of the code can just refer to that.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 3]
```diff
@@ -192,6 +243,23 @@ angular.module('kifi')
             return library.id === libraryId;
           });
         });
+      },
+
+      filterLibraries: function(term) {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:248]: Add space after `function`.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 4]
```diff
@@ -7,19 +7,38 @@ angular.module('kifi')
...
 
     var librarySummariesService = new Clutch(function () {
       return $http.get(routeService.getLibrarySummaries).then(function (res) {
-        util.replaceArrayInPlace(librarySummaries, res.data.libraries || []);
-        util.replaceArrayInPlace(invitedSummaries, res.data.invited || []);
+        var libs = res.data.libraries || [];
+        var invites = res.data.invited || [];
+        var i, lines;
+        for (i = 0; i < libs.length; i++) {
+          lines = shortenLibName(libs[i].name);
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:27]: This code can be DRYed up with the code in the next for-loop.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 5]
```diff
@@ -7,19 +7,38 @@ angular.module('kifi')
...
+    var fuseSearch2 = new Fuse(invitedSummaries, fuseOptions);
 
     var librarySummariesService = new Clutch(function () {
       return $http.get(routeService.getLibrarySummaries).then(function (res) {
-        util.replaceArrayInPlace(librarySummaries, res.data.libraries || []);
-        util.replaceArrayInPlace(invitedSummaries, res.data.invited || []);
+        var libs = res.data.libraries || [];
+        var invites = res.data.invited || [];
+        var i, lines;
+        for (i = 0; i < libs.length; i++) {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:26]: Try `forEach` for better clarity.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 6]
```diff
@@ -62,10 +81,42 @@ angular.module('kifi')
       return candidate;
     };
 
+    var maxLength = 22;
+    
+    function shortenLibName (fullName) {
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:86]: Remove space after `shortenLibName`.

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 7]
```diff
@@ -7,19 +7,38 @@ angular.module('kifi')
...
     var librarySummaries = [],
         invitedSummaries = [],
+        userLibsToShow = [],
+        invitedLibsToShow = [],
         libraryState = {}; // This is a variable to pass state from different library components of the system
-    /*
+    
     var fuseOptions = {
        keys: ['name'],
        threshold: 0.3
```
Comment (yipingliao) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:16]: Any explanation as to why the threshold is set to 0.3 and what that means?

**Comment by yipingliao** [comment]
Just a bunch of formatting nits and cleanup suggestions. The infinite loop definitely escaped me when I glanced at this yesterday. :(

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js (hunk 7 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:16]: Fuse allows for fuzzy search. Threshold is how far to go. Their docs are pretty sparse about meanings, but 0 means "exact match" and 1 means "any order of the letters"

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js [hunk 8]
```diff
@@ -7,19 +7,38 @@ angular.module('kifi')
...
     };
-    var fuseSearch = new Fuse(libraries, fuseOptions);
-    */
+    var fuseSearch1 = new Fuse(librarySummaries, fuseOptions);
```
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:18]: Better naming of this? `librarySummarySearch` and `invitedSummarySearch`?

File: kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js (hunk 6 above)
Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:86]: Is this the best way to do it? What about just using CSS?

Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:86]: +1 for CSS. Trailing `...` isn’t critical, in my opinion. Library names longer than two lines will probably be rare, and the users will likely already be familiar with them. If someone insists on a truncation indicator, we could use the second-line fade effect that we use in the extension.

Comment (andrewconner) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:86]: Multi-line ellipsis http://codepen.io/martinwolf/pen/qlFdp

Comment (2is10) [inline kifi-backend/modules/shoebox/angular/src/libraries/libraryService.js:86]: Cool, but that’s webkit-only, using an obsolete property.

The best CSS-only approach I’ve seen uses a fade and, optionally, an
ellipsis aligned to the box’s right edge. I find an ellipsis not aligned
with to the end of the text to be a bit gimmicky, but if it were really
faint, I could live with it.
http://www.cssmojo.com/line-clamp_for_non_webkit-based_browsers/

**Pull request merged**